docs/rfcs/036-sharded-ingest.md

```diff
@@ -0,0 +1,297 @@
+   require a reference to a Timeline.
+3. The logic that is current spread between `WalIngest` and `DatadirModification` for updating
+   metadata will be refactored to consume the metadata operations from the `ProcessedWalIngest`
+   and turn them into literal writes to metadata pages.  This part must be done sequentially.
+4. The resulting buffer of metadata page writes is combined with the buffer of relation block
+   writes, and written into the `InMemoryLayer`.
+
+### Phase 2: Decode & filter on safekeeper
+
+In the previous phase, the ingest code was modified to be able to do most of its work without access to
+a Timeline: this first stage of ingest simply converts a series of binary wal records into
+a buffer of relation/SLRU page writes, and a buffer of abstract metadata writes.
+
+The modified ingest code may be transplanted from pageserver to safekeeper (probably via a
+shared crate).  The safekeeper->pageserver network protocol is modified to:
+ - in subscription requests, send the `ShardIdentity` from the pageserver to the safekeeper
+ - in responses, transmit a `ProcessedWalIngest` instead of a raw `WalRecord`.
+ - use the `ShardIdentity` to filter the `ProcessedWalIngest` to relevant content for
+   the subscribing shard before transmitting it.
+
+The overall behavior of the pageserver->safekeeper interaction remains the same, in terms of
+consistent LSN feedback, and connection management.  Only the payload of the subscriptions
+changes, to express an LSN range of WAL as a filtered `ProcessedWalIngest` instead of the
+raw data.
+
+The ingest code on the pageserver can now skip the part where it does the first phase of
+processing, as it will receive pre-processed data off the wire.
+
+TODO: define batching on the safekeeper: will we do the same batching as on the pageserfer
+today?  (i.e. insteadof sending one ProcessedWalIngest per WalRecord, we would batch
+many WalRecords)
+
+### Phase 3: Shared safekeeper decode cursor for all shards
+
+In the previous phase, the initial processing of WAL was moved to the safekeeper, but it is still
+done once for each shard: this will generate O(N_shards) CPU work on the safekeeper.
+
+To avoid this, we will introduce a cursor concept for handling the WAL on the safekeeper: under
+normal operation, the WAL will be read from disk and decoded _once_, and the per-shard `ProcessedWalIngest` objects will be buffered, ready for the relevant shards to subscribe
+and receive them.
+
+The main challenge to implementing such cursor schemes is that pageserver shards may not all
+subscribe in a nicely synchronized way, and they may go backwards. We may make two useful simplifying
+assumptions:
+1. that there is little value in letting some shards proceed far ahead of others, since whichever
+   shard has the lowest ingested LSN is likely to act as a brake on postgres's progress the next
+   time it happens to request a page from that shard.
```

Comment:
There are a couple of risks to consider here:

* We may build up a large backlog of work, and only when the laggard catches up will we process it. This can contribute to longer recovery times and tail latencies. This is mitigated by backpressuring the compute, assuming the receive buffers have a reasonable bound.

* We're vulnerable to a thundering herd, where loss of one or more pageservers may block processing of a large number of timelines, and all of these will likely request catch-up at the same time. To avoid overwhelming the safekeeper we can use a semaphore to limit either the number of concurrent catchups or the amount of memory available for catchup, but this will increase tail recovery times.

I think this is likely a reasonable tradeoff, but it's worth testing it out under realistic conditions.